File: src/queue.cpp
#include "queue.h"

#include <cstring>
#include <fstream>
#include <iostream>

using namespace std;

void Queue::init() {
    front = nullptr;
    rear = nullptr;
}

void Queue::enqueue(const string& value) {
    QueueNode* newNode = new QueueNode{value, nullptr};
    if (rear != nullptr) {
        rear->next = newNode;
    }
    rear = newNode;
    if (front == nullptr) {
        front = rear;
    }
}

void Queue::dequeue() {
    if (front == nullptr) {
        return;
    }
    QueueNode* temp = front;
    front = front->next;
    if (front == nullptr) {
        rear = nullptr;
    }
    delete temp;
}

void Queue::print() {
    QueueNode* temp = front;
    while (temp != nullptr) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

void Queue::destroy() {
    while (front != nullptr) {
        dequeue();
    }
}

void Queue::loadFromFile(const string& fileName) {
    ifstream file(fileName);
    string value;
    while (file >> value) {
        enqueue(value);
    }
    file.close();
}

void Queue::saveToFile(const string& fileName) {
    ofstream file(fileName);
    QueueNode* temp = front;
    while (temp != nullptr) {
        file << temp->data << endl;
        temp = temp->next;
    }
    file.close();
}

void runQueue(int argc, char* argv[]) {
    Queue queue;
    queue.init();

    string fileName;
    string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        } else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    queue.loadFromFile(fileName);

    string command;
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        query = query.substr(pos + 1);
    } else {
        command = query;
    }

    if (command == "QPUSH") {
        queue.enqueue(query);
        queue.saveToFile(fileName);
    } else if (command == "QPOP") {
        queue.dequeue();
        queue.saveToFile(fileName);
    } else if (command == "PRINT") {
        queue.print();
    }

    queue.destroy();
}

File: Go/src/queue.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// QueueNode — узел очереди
type QueueNode struct {
	data string
	next *QueueNode
}

// Queue — очередь FIFO
type Queue struct {
	front *QueueNode
	rear  *QueueNode
}

func (q *Queue) Init() {
	q.front = nil
	q.rear = nil
}

func (q *Queue) Enqueue(value string) {
	newNode := &QueueNode{data: value, next: nil}
	if q.rear != nil {
		q.rear.next = newNode
	}
	q.rear = newNode
	if q.front == nil {
		q.front = q.rear
	}
}

func (q *Queue) Dequeue() {
	if q.front == nil {
		return
	}
	temp := q.front
	q.front = q.front.next
	if q.front == nil {
		q.rear = nil
	}
	_ = temp
}

func (q *Queue) Print() {
	first := true
	for cur := q.front; cur != nil; cur = cur.next {
		if !first {
			fmt.Print(" ")
		}
		first = false
		fmt.Print(cur.data)
	}
	fmt.Println()
}

func (q *Queue) Destroy() {
	q.front = nil
	q.rear = nil
}

func (q *Queue) LoadFromFile(fileName string) error {
	f, err := os.Open(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		q.Enqueue(scanner.Text())
	}
	return scanner.Err()
}

func (q *Queue) SaveToFile(fileName string) error {
	f, err := os.Create(fileName)
	if err != nil {
		return err
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for cur := q.front; cur != nil; cur = cur.next {
		_, _ = w.WriteString(cur.data + "\n")
	}
	return w.Flush()
}

func runQueue(fileName string, fullQuery string) {
	q := Queue{}
	q.Init()
	_ = q.LoadFromFile(fileName)

	parts := strings.SplitN(fullQuery, " ", 2)
	command := parts[0]
	arg := ""
	if len(parts) > 1 {
		arg = parts[1]
	}

	switch command {
	case "QPUSH":
		q.Enqueue(arg)
		_ = q.SaveToFile(fileName)
	case "QPOP":
		q.Dequeue()
		_ = q.SaveToFile(fileName)
	case "QPRINT":
		q.Print()
	default:
		fmt.Fprintln(os.Stderr, "Unknown queue command:", command)
	}

	q.Destroy()
}
